clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp

```diff
@@ -127,13 +127,26 @@ deduceDepTarget(const std::string &OutputFile,
 }
 
 /// Sanitize diagnostic options for dependency scan.
-static void sanitizeDiagOpts(DiagnosticOptions &DiagOpts) {
-  // Don't print 'X warnings and Y errors generated'.
-  DiagOpts.ShowCarets = false;
-  // Don't write out diagnostic file.
-  DiagOpts.DiagnosticSerializationFile.clear();
-  // Don't treat warnings as errors.
-  DiagOpts.Warnings.push_back("no-error");
+static void sanitizeDiagOpts(DiagnosticOptions &DiagOpts,
+                             ScanningOutputFormat ScanFormat) {
+  switch (ScanFormat) {
+  case ScanningOutputFormat::Make:
+  case ScanningOutputFormat::Full:
+  case ScanningOutputFormat::FullTree:
+    // Don't print 'X warnings and Y errors generated'.
+    DiagOpts.ShowCarets = false;
+    // Don't write out diagnostic file.
+    DiagOpts.DiagnosticSerializationFile.clear();
+    // Don't treat warnings as errors.
+    DiagOpts.Warnings.push_back("no-error");
+    break;
+  case ScanningOutputFormat::Tree:
+  case ScanningOutputFormat::IncludeTree:
```

Comment:
* Why is `Tree` affected?
* Is the idea here that the scanner will write to the serialized diagnostics file for the TU? That might be okay if scanning is implicit, but if the build system is running the scanner explicitly, shouldn't it detect failure through the API, in which case it wouldn't expect the scanner to touch the TU output files?

Reply:
> Why is Tree affected?

I can't test the include-tree with driver invocations yet, but, since `Tree` vs `IncludeTree` should be an implementation detail that is not observed by the build system, this change should be working fine for both anyway.

> shouldn't it detect failure through the API, in which case it wouldn't expect the scanner to touch the TU output files?

Are you saying that the build system may not want to write out a diagnostics file at all? Or that it does want to write out a diagnostics file but another mechanism should be used to write out the errors that the scanner emitted?

Reply:
I'm saying it's weird to run the scanner for a TU and have it implicitly write diagnostics to a file that is mentioned in that TU's command-line: the scanner wouldn't touch those files on success, so why would it on failure? You still need to know the scan failed via the API to even know if it would be correct to read that diag file or if it could be stale from a a previous build

Reply:
How should it work? I find it a bit weird that the interaction with the build system is that it assumes that the scanner never fails from diagnostic errors (all errors are ignored, even if the scanner doesn't find headers) and the errors are deferred for later when trying to run the compilation. Is this the intended interaction with the build system that we want?

Reply:
I'm not totally clear where errors are being ignored.`DependencyScanningWorker::computeDependencies`, which is the API used for "full" and "make" dependency scanning in clang-scan-deps, and is also the API used by the libclang scanning APIs on the "next" branch, returns an `Error` if there are any error diagnostics emitted, and prints the diagnostics into that Error's message.  I just checked, and that seems to work.  For `computeDependenciesFromCompilerInvocation` it has a `DiagsConsumer` .  If errors are ignored, it must be at a higher level, right?

The key for me is that the scanner APIs should not implicitly be writing to files - they should report failure in some explicit way controlled by the caller of the API.  For libclang, that could either be an error string (that's how it is currently), or I could see an argument to add a `CXDiagnosticSet` output.  For the main C++ scanner APIs, I think that should either by Error/Expected or a DiagnosticConsumer, which is already how it works AFAICT.  If you want the scanning daemon to write to the diagnostics files, that might be okay, but it should be explicit in the API.  A similar argument applies to .d files, which are currently controlled by the `EmitDependencyFile` flag.  Currently that flag is forced to false in `computeDependencies` and true in `computeDependenciesFromCompilerInvocation` but it should probably be a parameter controlled by the client.

CC @jansvoboda11 since this seems like a fundamental API question for the scanner.

Reply:
Per offline discussion, it is fine for the clang driver's implicit dep-scanning to be writing out a serialized diagnostics file, but it's not appropriate for explicit dep-scanning via libclang API to do that, (when using explicit libclang APIs the build system should not need the serialized diagnostics file in the first place), so essentially what needs to change for this PR is to provide an option about whether the dep-scanner needs to write to a serialized diagnostics file or not.

Reply:
I've added a boolean option `DiagGenerationAsCompilation` that indicates that diagnostic errors should emitted same way as the normal compilation, with "N errors generated" message and the serialized diagnostics file emitted if present in diagnostic options (it practically controls whether `sanitizeDiagOpts()` will get called or not).